core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/UnwrapCastInComparison.java

```diff
@@ -159,36 +163,153 @@ public Visitor(PlannerContext plannerContext, Session session)
+                            }
+                            greaterThanOrEqualLowBoundUnwrappedExpression = new Constant(sourceType, nextAfterLowBoundValue.get());
+                        }
+                        else {
+                            greaterThanOrEqualLowBoundUnwrappedExpression = lowBoundUnwrappedComparison.right();
+                        }
 
-            Comparison.Operator operator = expression.operator();
+                        Expression lessThanOrEqualHighBoundUnwrappedExpression;
+                        Optional<Object> previousBeforeHighBoundValue = Optional.empty();
+                        if (highBoundUnwrappedComparison.operator() == LESS_THAN) {
+                            previousBeforeHighBoundValue = sourceType.getPreviousValue(highBoundValue);
+                            if (previousBeforeHighBoundValue.isEmpty()) {
+                                return and(lowBoundUnwrappedComparison, highBoundUnwrappedComparison);
+                            }
+                            lessThanOrEqualHighBoundUnwrappedExpression = new Constant(sourceType, previousBeforeHighBoundValue.get());
+                        }
+                        else {
+                            lessThanOrEqualHighBoundUnwrappedExpression = highBoundUnwrappedComparison.right();
+                        }
+
+                        if (nextAfterLowBoundValue.isPresent() && previousBeforeHighBoundValue.isPresent()) {
+                            int compareNextLowAndPreviousHighBound = compare(sourceType, nextAfterLowBoundValue.get(), previousBeforeHighBoundValue.get());
+                            if (compareNextLowAndPreviousHighBound >= 0) {
+                                return falseIfNotNull(cast.expression());
+                            }
+                        }
+                        else if (previousBeforeHighBoundValue.isPresent()) {
+                            int compareLowAndPreviousHighBound = compare(sourceType, lowBoundValue, previousBeforeHighBoundValue.get());
+                            if (compareLowAndPreviousHighBound > 0) {
+                                return falseIfNotNull(cast.expression());
+                            }
+                        }
+                        else if (nextAfterLowBoundValue.isPresent()) {
+                            int compareNextLowAndHighBound = compare(sourceType, nextAfterLowBoundValue.get(), highBoundValue);
+                            if (compareNextLowAndHighBound > 0) {
+                                return falseIfNotNull(cast.expression());
+                            }
+                        }
+
+                        return new Between(cast.expression(), greaterThanOrEqualLowBoundUnwrappedExpression, lessThanOrEqualHighBoundUnwrappedExpression);
+                    }
+                }
+
+                return and(lowBoundUnwrappedComparison, highBoundUnwrappedComparison);
+            }
+
+            return expression;
```

Comment:
The `unwrapCast` method might return constant `FALSE`. I don't think this case is covered. The result would be `FALSE`.